### PR TITLE
Improve threadpool usage and error handling for API key validation

### LINKED ISSUE #58088: Avoid saturating GET thread pool with API key hash verification
Currently `ApiKeyService#validateApiKeyCredentials` is performed in the same GET thread which fetches the API key document. When the API key is not cached, the PBKDF2 10,000 (default) hashing is required and it takes a long time to finish (~70 ms). This leads to saturation of the GET thread pool when a large number of clients try to connect for the first time. 

When the pool is saturated by a large number of new clients, it not only blocks the new clients but also blocks the existing clients. This is because all clients require fetching their API key document through the GET thread pool. This leads to instability for all clients. It does _not_ feel right to block existing clients from indexing just because there are new clients trying to connect. A possible approach is to have a dedicated thread pool for secret hashing work. Ideally this pool should be shared for all hashing work, e.g. native realm etc. In practice, the first step would be creating it for ApiKeyService to offload the computation cost from the GET thread pool.

Another issue is the response status when the pool is saturated. It currently returns `401` because we do not want leak any information to unauthenticated clients. However, this also makes it impossible for a legitimate client to tell whether they should retry on failures. If the server is too busy to handle authentication, return something like `429` could be more appropriate. Since authentication is _not_ even attempted, this information (server is busy) is likely safe to share.

### PR DESCRIPTION
**Pull request by ywangd**
The PR introduces following two changes:
* Move API key validation into a new separate threadpool 
* Return more informative response on threadpool saturation

The new threadpool is created separately with half of the available processors and 1000 in queue size. We could combine it with the existing TokenService's threadpool. Technically it is straightforward, but I am not sure whether it could be a rushed optimization since I am not clear about potential impact on the token service.

On threadpoool saturation, it now fails with `EsRejectedExecutionException` which in turns gives back a `429` status code to users. ~Note this is also a subtle behaviour change: Previously any failures during API key validation are translated into "_unsuccessful but continue to realm authentication_". After the change, threadpool saturation error is translated into "_unsuccessful and terminate authentication_". The difference will manifest iteself when user sends in two set of credentials, e.g. one for API key and one for basic auth. Before the change, authentication will continue with the basic auth and if it is valid, authentication will end up as successful. After the change, the authentication stops at API key when pool is saturated and does not proceed further. When threadpool is saturated, it is highly likely that users do want the API key authentication (otherwise the pool will not be saturated in the first place). Hence I doubt any user would really depend on the existing behaviour.~ (edit: this is not a concern since the code does not allow multiple Authorization headers. Thanks @jkakavas)

Resolves: #58088

### CODE REVIEW

**Comment by ywangd** [comment]
Just realised that shunting everything after GetDoc call (mainly `ApiKeyService#validateApiKeyCredentials`) to a new thread pool does not solve the problem that cached API key auth is blocked by uncached API key auth, i.e. existing indexing operations are made unstable because new clients try to connect. 

So I made changes to [only push `ApiKeyService#verifyKeyAgainstHash`](https://github.com/elastic/elasticsearch/pull/58090/files#diff-cc4f9d4c564aae71214ac709ed789b4dR580-R586) to the new thread pool and leave cached API key auth to the same GET thread pool. Accordingly to the performance tests, once all API keys are cached, the GET thread pool can sustain at least 5000 auth requests per second. So this change seems to be the best of both worlds.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 1]
```diff
@@ -466,25 +480,33 @@ void validateApiKeyCredentials(String docId, Map<String, Object> source, ApiKeyC
                                 validateApiKeyCredentials(docId, source, credentials, clock, listener);
                             }
                         }, listener::onFailure),
-                        threadPool.generic(), threadPool.getThreadContext());
+                        threadPool.executor(THREAD_POOL_NAME), threadPool.getThreadContext());
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:483]: The specified thread pool will only be used when the future is being computed, i.e. expensive hashing is underway. In this case, the API key is not yet cached, so change it to use the separate hashing pool which are dedicated to uncached API keys.

**Comment by jkakavas** [comment]
> We could combine it with the existing TokenService's threadpool. Technically it is straightforward, but I am not sure whether it could be a rushed optimization since I am not clear about potential impact on the token service.

Agreed, I don't see any obvious benefit for this or solving any actual problem we have right now, but I can see a clear negative impact on token based authentication which we should avoid. 

As discussed in slack, I'm not particularly worried about multiple `Authorization` headers as we only handle the first one right now

**Review by albertzaharovits — COMMENTED** [review]
Overall this looks good!
I have raised two topics for discussion:
- I have a suggestion about what I believe is a better place for handling EsRejectedExecutionException 
- I think it's better to enqueue only expensive hashing on the new thread pool, and not all (most) API verifications.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 2]
```diff
@@ -334,8 +342,14 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
                         AuthenticationResult.unsuccessful("unable to find apikey with id " + credentials.getId(), null));
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:342]: You are right. It is better to confine the special handling in a single method. Updated as you suggested with one modification: instead of calling `listener.onResponse(AuthenticationResult.unsuccessful(...))` on the `else` branch, I think it is better to call `listener.onFailure(e)` which is consistent with original logic. We might wanna change how `unsuccessful` and `terminate` are determined in the future, but I feel it deserves a separate PR.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 3]
```diff
@@ -307,7 +311,11 @@ void authenticateWithApiKeyIfPresent(ThreadContext ctx, ActionListener<Authentic
                     },
                     e -> {
                         credentials.close();
-                        listener.onFailure(e);
+                        if (ExceptionsHelper.unwrapCause(e) instanceof EsRejectedExecutionException) {
+                            listener.onResponse(AuthenticationResult.terminate("server is too busy to respond", e));
+                        } else {
+                            listener.onFailure(e);
+                        }
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: Agreed. See above. Thanks

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 4]
```diff
@@ -552,14 +574,17 @@ static ApiKeyCredentials getCredentialsFromHeader(ThreadContext threadContext) {
     }
 
     // Protected instance method so this can be mocked
-    protected boolean verifyKeyAgainstHash(String apiKeyHash, ApiKeyCredentials credentials) {
+    protected void verifyKeyAgainstHash(String apiKeyHash, ApiKeyCredentials credentials, ActionListener<Boolean> listener) {
         final char[] apiKeyHashChars = apiKeyHash.toCharArray();
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: I think this is a subtle but great suggestion. We want minimize the duration a secret stay in memory. When it is outside of the runnable, it could stay in memory for very long if thread pool is busy. Moved it into the runnable as you suggested.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: Reverted it back to use generic pool. I don't know exactly the best practice of using generic pool. I think what you said makes sense. But on the other hand, generic pool being unbounded is likely an issue by itself. It is likely a bad circuit breaker. Anyway, we have been using generic pool here for a while now, so it is safe to keep it untouched.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 5]
```diff
@@ -552,14 +574,17 @@ static ApiKeyCredentials getCredentialsFromHeader(ThreadContext threadContext) {
...
-        try {
-            Hasher hasher = Hasher.resolveFromHash(apiKeyHash.toCharArray());
-            return hasher.verify(credentials.getKey(), apiKeyHashChars);
-        } finally {
-            Arrays.fill(apiKeyHashChars, (char) 0);
-        }
+        Hasher hasher = Hasher.resolveFromHash(apiKeyHash.toCharArray());
+        final String executorName = Thread.currentThread().getName().contains(THREAD_POOL_NAME) ? ThreadPool.Names.SAME : THREAD_POOL_NAME;
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: Thanks. Removed since the generic pool usage is now restored.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 6]
```diff
@@ -992,9 +993,14 @@ public AsyncSender interceptSender(AsyncSender sender) {
...
         if (enabled) {
-            return Collections.singletonList(
-                    new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000, "xpack.security.authc.token.thread_pool",
-                        false));
+            final int allocatedProcessors = EsExecutors.allocatedProcessors(settings);
+            return List.of(
+                new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000,
+                    "xpack.security.authc.token.thread_pool", false),
+                new FixedExecutorBuilder(settings, ApiKeyService.THREAD_POOL_NAME,
+                    (allocatedProcessors + 1) / 2, 1000,
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1001]: Note by using half of the allocated processors, initial authentiation of API keys, i.e. warming up the cache, could be up to twice as slow as current implemention.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:344]: > instead of calling listener.onResponse(AuthenticationResult.unsuccessful(...)) on the else branch, I think it is better to call listener.onFailure(e) which is consistent with original logic

Great, it was my intention to maintain the original behavior.

> We might wanna change how unsuccessful and terminate are determined in the future, but I feel it deserves a separate PR.

Yeah, I believe we should do that.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java [hunk 7]
```diff
@@ -602,6 +610,83 @@ public void testWillGetLookedUpByRealmTypeIfExists() {
...
+        final ApiKeyCredentials creds = new ApiKeyCredentials(randomAlphaOfLength(12), new SecureString(apiKey.toCharArray()));
+        writeCredentialsToThreadContext(creds);
+
+        Hasher hasher = randomFrom(Hasher.PBKDF2, Hasher.BCRYPT4, Hasher.BCRYPT);
+        final char[] hash = hasher.hash(new SecureString(apiKey.toCharArray()));
+        Map<String, Object> sourceMap = buildApiKeySourceDoc(hash);
+        mockSourceDocument(creds.getId(), sourceMap);
+        final ExecutorService mockExecutorService = mock(ExecutorService.class);
+        when(threadPool.executor(ApiKeyService.THREAD_POOL_NAME)).thenReturn(mockExecutorService);
+        Mockito.doThrow(new EsRejectedExecutionException("rejected")).when(mockExecutorService).execute(any(Runnable.class));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java:637]: This is not how the `EsThreadPoolExecutor` [handles over-capacity](https://github.com/elastic/elasticsearch/blob/09ff747fe7479f9c695725ec642315cf0343a533/server/src/main/java/org/elasticsearch/common/util/concurrent/EsThreadPoolExecutor.java#L84). The test below as well.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 6 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1001]: The **maximum theoretical throughput** of new (previously unseen) API key validations will indeed halve, but I'm not personally worried about it. We're talking about the theoretical maximum throughput which, because of the contention on the get threadpool, and because there are many thread pools overbooking the available processors, it's not something I would consider practically important.

Yet, a possible mitigation would be to decrease (halve) the default hashing cost factor for API Keys. GIven the length of the random api key secret that we generate, we would still remain out of the brute force plausibility range.

**Review by albertzaharovits — APPROVED** [review]
LGTM ! besides a minor thing in tests, this is ready to 🚢 
Great job Yang!

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java (hunk 7 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java]: Thanks for the great attention to details! 👍  It's now corrected.

**Comment by jaymode** [comment]
Was any consideration given to having a single thread pool for hashing/expensive operations within security?

**Comment by ywangd** [comment]
> Was any consideration given to having a single thread pool for hashing/expensive operations within security?

I thought about it but didn't have an open discussion within the team. My concern was the initial authentication would be too slow, but this could also be a bias that I got from doing performance tests where initial warming up is evident. It might not be as a big concern for real use cases? Do you have an argument in favor of it?

**Comment by jaymode** [comment]
> It might not be as a big concern for real use cases? Do you have an argument in favor of it?

In general, we have been very judicious with adding additional thread pools. This PR targets API keys but we also have similar issues with native users since we use a get request and then validate on the get thread pool there as well. Some hash verification can also occur on network threads in the case of the file realm. The reserved realm also will perform the hash verification on a network thread or thread from the get thread pool depending on the state of the security index. These operations are designed to take time so if we were going to introduce a new thread pool, I think it would be wise to use a single thread pool to move these expensive operations out of the other pools and limit the amount of CPU that could be scheduled to service these operations.

**Review by tvernum — CHANGES_REQUESTED** [review]
I'd like to see a test somewhere for the `429` response code behaviour.

We have a test for the ApiKeyService, but nothing that shows that `EsRejectedExecutionException` bubbles up into a `429` response. I think we need that.

**Comment by tvernum** [comment]
> >  Was any consideration given to having a single thread pool for hashing/expensive operations within security?
> I thought about it but didn't have an open discussion within the team

I recall having exactly that conversation and thought that we agreed to move in that direction (not in this PR, but as a followup).

**Comment by ywangd** [comment]
> > > Was any consideration given to having a single thread pool for hashing/expensive operations within security?
> 
> > I thought about it but didn't have an open discussion within the team
> 
> I recall having exactly that conversation and thought that we agreed to move in that direction (not in this PR, but as a followup).

Sorry @jaymode  I mis-read your message. I thought you were asking for a single **thread" thread pool .... Tim is right. We have discussed this and agreed that we will consolidate security related expensive operations into a single pool.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 6 above)
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1001]: That's quite a large number of new threads to be adding to the system, especially threads that could end up all being busy at once for a long time authenticating a lot of clients. They're then stealing CPU time from other components of the system. What work have we done to justify that the thread pool needs to be this large (the ceiling of half the number of cores)?

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1001]: A few reasons:
* It has fewer threads than the `GET` thread pool - the current code performs the hashing in the `GET` thread pool, which has number of threads equals to number of cores. The new thread pool should in fact help with resource stealing. Besides, saturation of GET thread pool has wider negative impacts to both uncached and cached authentication. A new pool helps isolate the impact to only new clients trying to connect.
* Balance between throughput and resource contention - the expensive hashing operations are unavoidable. Performance tests show about 5000 authentications per minute (8 cores). With half of the threads, the number will be halved as well. Assuming a 25K clients (50K keys) use case, it will take 20 min for all keys to be cached. Reducing the number of threads too much would push the warm up time even higher. Therefore using half number of the cores feels like a reasonable middle ground.
* Hashing in the `generic` thread pool (which uses all cores) does not have obvious performance hit - I tried simply pushing the hashing to the `generic` thread pool, which then used all cores for the hashing. I didn't observe any obivious performance issue with it. In fact, the initial warm up stage is smoother and has a better throughput than using the `GET` thread pool. Therefore it feels that using half of the cores should be relatively safe and also performant. I do plan to run a few more rounds of performance tests after all scheduled `v7.9` API key improvements. So we can be confident on this.
* The new thread pool will be shared by all security related expensive operations - currently its only usage is for hashing of API key auth, so it is not a very good reason. But it could soon see another use case: the API key creation also requires expensive hashing which currently runs in the `transport_worker` thread. I need to do some more investigation, but the result is very likely that we need move it to this new thread pool as well.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1001]: My _thoughts_  for approving the PR were that we operate under some desired rough approximation for the rate of authentication using keys (what Yang is describing on the second point above) and so by switching to a new pool (given that the get operation is so much faster than the hash, and that the new pool is smaller than the get pool) all the threads in the new pool will be busy at peak rate, so that we can estimate that peak rate (assuming a relatively tranquil system where the pool threads stay running). The original peak rate (using the get pool) was noisy because any get requests in the system interfered with it (in addition to the obvious flip side that authn hashing interfered with get requests in the system).

When the system is busy and all threads are working at the pool queues and no queue is empty, it all boils down to the processors' queue. In a sense all thread pool queues "merge" into processor queues, with the priority given by the relative number of threads ascribed to each pool. So introducing a new pool will decrease the throughput of existing pools, proportionate to relative sizes, in exchange of a guaranteed minimum throughput on the new pool ,because the new queue is not populated with any other types of work (compared to reusing an existing pool that contains other work types).

To summarise, I would generally introduce a thread pool when I want more control over the rate of some work (both the minimum and the maximum rate).

That's my thought process about thread pool queues upon which I've approved the PR. I believe a new thread pool is the right decision, but I concede that we've informed the sizing decision on rough external authn rates without also consulting with folks that tune the existing thread pools.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 8]
```diff
@@ -992,9 +993,14 @@ public AsyncSender interceptSender(AsyncSender sender) {
...
     public List<ExecutorBuilder<?>> getExecutorBuilders(final Settings settings) {
         if (enabled) {
-            return Collections.singletonList(
-                    new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000, "xpack.security.authc.token.thread_pool",
-                        false));
+            final int allocatedProcessors = EsExecutors.allocatedProcessors(settings);
+            return List.of(
+                new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000,
+                    "xpack.security.authc.token.thread_pool", false),
+                new FixedExecutorBuilder(settings, ApiKeyService.THREAD_POOL_NAME,
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: As discussed, renamed it to `security-crypto`.

**Comment by ywangd** [comment]
@tvernum Thank you for your suggestion, a [test](https://github.com/elastic/elasticsearch/pull/58090/files#diff-f1e82384e2498269e87f0e84280c8dbcR836) for the `429` status code is now added.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 6 above)
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1008]: Okay, thanks for the great explanations to help me understand the perspective here. One more question: is it enough to have a simple "security" thread pool? Do we need a token and a crypto thread pool and maybe another thread pool for other security aspects in the future?

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1008]: **TL;DR**: We are not considering another thread pool for other security aspects. 

Our intention for the new thread pool is to focus on "cpu-intensive" security computations, namely hashing and encryption, especially in the scope of authentication. Given their expensiveness and importance, we'd like to control their impact both ways: they should not take over all system resources but in the mean time should maintain a reasonable level of throughput. We use following criteria to justify whether an operation should be added to this pool:
* _Is it expensive_? The standard here is somewhat subjective. Based on performance tests, I personally would eye on anything longer than 1 ms.
* _Is it authentication/authorization related_? We'd like to prioritize these operations since they usually need to complete in a timely manner. 
* _Does it need to scale? i.e. potential burst in volume_? When there is a large increase in concurrency, we'd like to protect the system from being completely flooded. 

With above criteria, we can come up with the followings:
* Hashing for uncached results are great candidates. These include API Key auth (current PR, v7.9), API Key creation (v7.9?), tokens, username/password (planned, v7.10) 
* Signing and cryption for SAML assertions and odic claims are potential candidates (pending further performance check)
* Hashing for cached result is _not_ qualified for the new thread pool since it is fast (sha256, ~3 microseconds). 
* Manipulations of security documents are _not_ qualified either
* Encrypted snapshot, though potentially expensive, is not auth related and unlikely needs to scale. Hence it is also _not_ qualified.
* An interesting case is read/write of the security index. When the index is not on the local node, these operations can be expensive (above 1 ms) and it also matches the two other criteria. However the system indices work (#57936) is supposed to cover it. Hence we are _not_ considering it either.
* Any other security aspects that are _not_ qualified

Now the question is: _whether above _unqualified_ security operations worth to have their own separate thread pool?_

When security is enabled, security related operations are ubiquitous, e.g. every request needs to go through authentication and authorization flow. Currently our *general practice* is to "just let them run in the same thread", which could be `get`, `transport_worker` etc. This strategy has served us reasonally well so far. But with the upcoming change to expose the service to unprecedented large number of clients, we started noticing some issues revealed by the performance tests, which in turn leads to this PR and a few other work in the pipeline. At this stage, based on what we know, we still believe the existing strategy works fine for security operations that are _not_ qualified for the new thread pool. Hence we do _not_ plan to add another one. 

For the sake of completeness, there are exceptions to the existing strategy. Outbound LDAP calls are [submitted via the generic pool](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapUtils.java#L95-L97) to avoid risks of deadlocking if they are submitted via the same pool used by the outbound requests. There is no sign for this being an issue so far. We had brief discussion about it and decided to evaluate it more closely at a later time.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java [hunk 9]
```diff
@@ -823,6 +833,60 @@ public void testDerivedKeys() throws ExecutionException, InterruptedException {
...
+            executorService.submit(() -> {
+                readyLatch.countDown();
+                try {
+                    blockingLatch.await();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+        IntStream.range(0, 1000).forEach(i -> executorService.submit(() -> {}));
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java:873]: This `1000` is based on the `SECURITY_CRYPTO_THREAD_POOL_NAME` queue size, correct?

Can we extract it to a variable so that its meaning is clear?
```suggestion
        // Fill the whole queue for the crypto thread pool
        final int queueSize = 1000;
        IntStream.range(0, queueSize).forEach(i -> executorService.submit(() -> {}));
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java [hunk 10]
```diff
@@ -823,6 +833,60 @@ public void testDerivedKeys() throws ExecutionException, InterruptedException {
...
+
+        try (RestClient restClient = createRestClient(nodeInfos, null, "http")) {
+            final String base64ApiKeyKeyValue = Base64.getEncoder().encodeToString(
+                (createApiKeyResponse.getId() + ":" + createApiKeyResponse.getKey().toString()).getBytes(StandardCharsets.UTF_8));
+
+            final Request authRequest = new Request("GET", "_security/_authenticate");
+            authRequest.setOptions(RequestOptions.DEFAULT.toBuilder().addHeader(
+                "Authorization", "ApiKey " + base64ApiKeyKeyValue).build());
+            final ResponseException responseException = expectThrows(ResponseException.class, () -> restClient.performRequest(authRequest));
+            assertThat(responseException.getMessage(), containsString("429 Too Many Requests"));
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java:884]:
```suggestion
            assertThat(responseException.getMessage(), containsString("429 Too Many Requests"));
            assertThat(responseException.getResponse().getStatusLine().getStatusCode(), is(429));
```

**Review by tvernum — APPROVED** [review]
LGTM

**Review by jasontedor — APPROVED** [review]
LGTM.
